### ISSUE TITLE
Race condition caused ArrayIndexOutOfBoundsException on DispatcherSupplier.get()?

### ISSUE THREAD

**Issue by josebarrueta** [reactor/reactor#487]
Hi,

While running load tests against our staging environment (REST service on Spring) we started noticing ArrayIndexOutOfBoundsException in our logs, we configured a `dispatcherGroup` with pool size of `64`, after some time of running our tests, the exceptions appeared (and of course it never recovered).

From the logs file we get these messages (this is the first one).

```
2015-06-04T19:55:18.403+0000 ERROR [http-bio-8080-exec-506] c.s.i.w.s.h.RestExceptionHandler Unhandled exception occurred: 65
java.lang.ArrayIndexOutOfBoundsException: 65
	at reactor.Environment$2.get(Environment.java:911) ~[reactor-core-2.0.3.RELEASE.jar:na]
	at reactor.Environment$2.get(Environment.java:843) ~[reactor-core-2.0.3.RELEASE.jar:na]
	at com.stormpath.iam.web.rest.DefaultRestResourceFactory.lambda$convertCollectionResult$5(DefaultRestResourceFactory.java:279) ~[DefaultRestResourceFactory.class:na]
	at com.stormpath.iam.web.rest.DefaultRestResourceFactory$$Lambda$18/1861024664.apply(Unknown Source) ~[na:na]
	at reactor.rx.action.transformation.MapAction.doNext(MapAction.java:37) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.action.Action.onNext(Action.java:202) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.subscription.PushSubscription.onNext(PushSubscription.java:111) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.stream.IterableStream$1.request(IterableStream.java:72) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.action.Action.requestMore(Action.java:619) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.action.Action.requestUpstream(Action.java:572) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.action.Action$7.onRequest(Action.java:564) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.subscription.PushSubscription.start(PushSubscription.java:145) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.action.Action.doStart(Action.java:182) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.action.Action.onSubscribe(Action.java:172) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.stream.IterableStream.subscribe(IterableStream.java:64) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.stream.LiftStream.subscribe(LiftStream.java:81) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.stream.LiftStream.subscribe(LiftStream.java:81) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.stream.LiftStream.subscribe(LiftStream.java:81) ~[reactor-stream-2.0.3.RELEASE.jar:na]
	at reactor.rx.Stream.next(Stream.java:2800) ~[reactor-stream-2.0.3.RELEASE.jar:na]
```

The Stream is really simple it looks like:

```java 
           
            AtomicInteger entityIndex = new AtomicInteger();

            try {
                List<Tuple2<Integer, Map>> resultTuples = Streams.from(resultItems).flatMap(
                    m -> Streams.just(Tuple.of(entityIndex.getAndIncrement(), m))
                                .dispatchOn(dispatcherSupplier.get())
                                .map(input -> entityTransformer.apply(requestContext, input)))
                                                                 .buffer(resultItemsSize).next()
                                                                 .await(serverResponseTimeout, TimeUnit.MILLISECONDS);


                Assert.state(resultTuples != null && resultTuples.size() == resultItemsSize,
                             "The resolved result items don't match the size of the input.");

                Map[] tempArray = new Map[resultItemsSize];
                for (Tuple2<Integer, Map> tuple : resultTuples) {
                    tempArray[tuple.getT1()] = tuple.getT2();
                }
                convertedItems = Arrays.asList(tempArray);

            } catch (InterruptedException e) {
                .....
            }
```

The DispatcherSupplier is a Spring Bean configured with a pool size of 64

```
   @Bean
    public DispatcherSupplier dispatcherSupplier() {
        return environment.getCachedDispatchers();
    }
```

Checking the code from the logs, the failing part looks like the `DispatcherSupplier#getNextIndex()` at some point returns an invalid one, probably due to a race condition of Two threads updating the `roundRobinIndex` value.

**Comment by jbrisbin** [comment]
Looks like we might have an `==` where we intend a `>=` in the round robin code.

**Comment by josebarrueta** [comment]
Yeah that's what I thought too, but event with that there is the potential of failing once right?

**Comment by jbrisbin** [comment]
We probably should take a closer look at how we're doing that for a 2.0.4 release. That's a very critical section of code and we need to be pretty careful about what we do there since even the slightest unnecessary overhead could have detrimental effects to throughput.

**Comment by josebarrueta** [comment]
Any recommendations in the meantime? I was thinking to move to a `workDispatcher` in the meantime and do `subscribeOn()` instead of dispatch?

**Comment by sofiaguyang** [comment]
Any updates on this? I am also encountering this problem during my load testing. I am using the 2.0.6.RELEASE version as bundled in Grails 3.0.12.

**Comment by smaldini** [comment]
Can you force the dependency and try with 2.0.7.RELEASE ?

**Comment by sofiaguyang** [comment]
I've forced dependencies to all use 2.0.7.RELEASE. I ran my load test a couple of times and did not encounter the ArrayIndexOutOfBoundsException problem -- unsure though if it's just luck that no race condition occurred, or it's truly been fixed already. I compared the Environment classes of 2.0.6.RELEASE and 2.0.7.RELEASE and they appear to be identical.

**Comment by smaldini** [comment]
@sofiaguyang ping if you can try 2.0.8.BUILD-SNAPSHOT

**Comment by sofiaguyang** [comment]
@smaldini,

Pulled 2.0.8.BUILD-SNAPSHOT as dependency. I ran my load test which runs 100 concurrent requests at any given time and I still get ArrayIndexOutOfBoundsException. 

<img width="1413" alt="screen shot 2016-02-06 at 12 27 50 am" src="https://cloud.githubusercontent.com/assets/4891894/12852166/96b371d2-cc68-11e5-9816-5fcc04c0dacd.png">

**Comment by smaldini** [comment]
by bad if you can retry with a refreshed snapshot.

**Comment by sofiaguyang** [comment]
I refreshed the snapshot -- but content for the Environment.java file seem to be identical?

**Comment by smaldini** [comment]
Snapshot is stuck can you clone reactor , checkout 2.0.x branch and do a ./gradlew install to force a local cache refresh ?

**Comment by akarnokd** [comment]
This index handling looks odd:
```java
		int getNextIndex() {
			int index;
			for (; ; ) {
				index = this.index.get() + 1;
				if (index == poolsize) {
					if (this.index.compareAndSet(index, 0)) {
						return 0;
					}
				}
				else {
					return index;
				}
			}
		}
```

At the start, the index is -1, the first caller returns 0, the second caller returns 0, etc. and never seem to do any round robin. If `poolsize` is zero, then there is a CAS from -1 to 0 and 0 is returned which is clearly going to cause the index out of bounds error.

So I'd check if poolsize > 0 in your environment so at least you have access to one dispatcher.

**Comment by smaldini** [comment]
Indeed thanks David I have overlooked the issue and I am really to blame here but I'll fix that as soon as I get back home today.

**Comment by sofiaguyang** [comment]
@smaldini, I cloned reactor and installed your latest fix. I am now getting the following error: 
<img width="817" alt="screen shot 2016-02-09 at 8 01 09 pm" src="https://cloud.githubusercontent.com/assets/4891894/12915790/75f9a27e-cf68-11e5-9c29-80830adda0c8.png">

The load test is configured to fire 10,000 requests through 100 simultaneous connections. I am unsure whether this is still an issue with reactor or now an issue with Grails. I also noticed that the load-test seemed to take longer now before the error crept in.

**Comment by sofiaguyang** [comment]
@smaldini, please disregard my last reply. Turns out that it's a false alarm and I found the problem was with my REST call to an external service which was not responding. 

So having resolved that lapse, I've tried running my benchmark test at least 10 times and have not encountered the ArrayIndexOutOfBoundsException. 

Think we can close this issue now :+1:

**Comment by smaldini** [comment]
I will release 2.0.8.RELEASE alongside 2.5.0.M1 for Stream then.

**Comment by hdlopez** [comment]
sorry guys, I know this issue is closed, but I have the same issue with grails 3.0.1, could you explain me in a few words, how I could force the dependency and try with 2.0.8.RELEASE ? 

Maybe it's a silly question and it's just sth like "force()" ha

I have already added all the dependencies in my gradle file, I refreshed them, but I'm still seeing the 2.0.0.RELEASE dependency 

Thanks in advance for you help
